tests/deploy_test.go

```diff
@@ -129,6 +135,51 @@ func Test_Deploy_MetaData(t *testing.T) {
 			}
 		})
 	}
+
+	t.Run("test listing functions", func(t *testing.T) {
+
+		listCases := map[string]map[string]types.FunctionDeployment{}
+		for _, tc := range cases {
+			fncs, ok := listCases[tc.function.Namespace]
+			if !ok {
+				fncs = map[string]types.FunctionDeployment{}
+			}
+
+			fncs[tc.function.Service] = tc.function
+			listCases[tc.function.Namespace] = fncs
+		}
+
+		for namespace, expected := range listCases {
+			actual, err := config.Client.ListFunctions(ctx, namespace)
+			require.NoError(t, err, "unable to List function in namspace")
+
+			require.Equal(t, len(expected), len(actual), "unexpected number of functions")
+
+			for _, actualF := range actual {
+				expectedF, ok := expected[actualF.Name]
+				require.True(t, ok, "unexpected function found")
+				require.Equal(t, expectedF.Service, actualF.Name)
+				require.Equal(t, expectedF.Image, actualF.Image, "incorrect image")
+				require.Equal(t, expectedF.Namespace, actualF.Namespace, "incorrect namespace")
+				require.Equal(t, expectedF.EnvProcess, actualF.EnvProcess, "incorrect env process")
+				require.Equal(t, expectedF.EnvVars, actualF.EnvVars, "incorrect environment variables")
+				require.Equal(t, expectedF.Constraints, actualF.Constraints, "incorrect constraints")
+				require.Equal(t, expectedF.Secrets, actualF.Secrets, "incorrect secrets")
+				require.Equal(t, expectedF.Limits, actualF.Limits, "incorrect function limits")
+				require.Equal(t, expectedF.Requests, actualF.Requests, "incorrect function requests")
+				require.Equal(t, expectedF.ReadOnlyRootFilesystem, actualF.ReadOnlyRootFilesystem, "incorrect read only filesystem flag")
+
+				// we expect all systems to add the `faas_function` label?
+				expectedLabels := copyStrMap(expectedF.Labels)
+				expectedLabels["faas_function"] = expectedF.Service
+				require.Equal(t, &expectedLabels, actualF.Labels, "incorrect labels")
+
+				// some systems add additional annotations, we remove those
+				actualAnnotations := removeSystemFields(actualF.Annotations)
+				require.Equal(t, expectedF.Annotations, &actualAnnotations, "incorrect annotations")
```

Comment:
@alexellis i know we haven't used this package yet, but what do you think? these tests are _much_ easier to write and read because the equality logic is well handled by the `require` package. Otherwise we need to write and maintain _a lot_ more code to get the same results.

Reply:
The increase in coverage is good to see

I think the test is large, why is that? Are we testing too many things?

We may do better with using a comparison or serialisation and string compare.

I also think utility classes may be useful here. 

Assert/testify is still not idiomatic Go, not keen on it being introduced here, but would like to hear alternative suggestions.

Reply:
I don't understand what you mean by the test being large? All it does is verify that the  List response is actually complete and matches the values that were used in the Deployment. The response has a lot of fields so the test is longer. What are you expecting it to _not_ test?

Regarding testify vs utility functions, I am not sure i understand why one is idiomatic and the other isn'tt. The alternative, to write utility test functions, ends up looking the same: [`listContains`](https://github.com/openfaas/certifier/blob/fd3eada327e66c71e715d997b8f4edde4a1c1eb6/tests/secretCRUD_test.go#L97-L104) vs `require.Contains`.  Except one is hand rolled vs a well tested and widely used library.

Comparing via serialization is a huge pain in the ass, the tests I have at work that depend on serialization comparisons tend to be the most fragile and the hardest to actually update and prone to silly bugs due to typos. When we change the sdk it is also harder to refactor because we don't actually reference sdk fields or types, instead it is just a lot of json strings, the linter/compiler are completely useless then.  When tests based on serialization break they are also harder to understand _why_ because you have a very long json string and some part of it is different  but it isn't obvious _which_ part is different. And depending on how you write it, it can be sensitive to whitespace or ordering in a non-meaningful way.t  If we really want to compare serializations, then we shouldn't use the `sdk` at all because it hides the actual response from the provider.

Reply:
https://blog.alexellis.io/golang-writing-unit-tests/

> You may be more used to using the Assert keyword to perform checking, but the authors of The Go Programming Language make some good arguments for Go's style over Assertions.

https://blog.alexellis.io/golang-writing-unit-tests/

![stretch](https://user-images.githubusercontent.com/6358735/118011939-7660ba80-b348-11eb-9b46-f2acdf229c5f.png)

Reply:
Could something like this be used? https://github.com/stretchr/testify/blob/acba37e5db06f0093b465a7d47822bf13644b66c/assert/assertions.go#L80

Reply:
Where we did this in faas-netes: https://github.com/openfaas/faas-netes/blob/e5b3e8af8d9b91d114abed1d61bdfd15bfcefc85/pkg/controller/deployment.go#L272

Reply:
@alexellis i removed it, nothing but std lib now, hopefully it looks ok now